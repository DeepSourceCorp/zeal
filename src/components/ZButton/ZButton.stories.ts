import ZButton from './ZButton.vue'

import { action } from '@storybook/addon-actions'

import '../../assets/css/tailwind.css'
import '../../assets/css/typography.css'
import '../../assets/css/layout.css'

export default {
  title: 'Zeal/Button',
  component: ZButton,
  excludeStories: /.*Data$/
}

export const DefaultWithoutText = () => ({
  components: { ZButton },
  template: '<z-button></z-button>'
})

export const ButtonWithText = () => ({
  components: { ZButton },
  template: '<z-button color="primary">Hello World</z-button>'
})

export const Colors = () => ({
  components: { ZButton },
  template: `<div>
        <z-button color="primary">Hello World</z-button>
        <z-button color="secondary">Hello World</z-button>
    </div>`
})

export const Sizes = () => ({
  components: { ZButton },
  template: `<div>
        <z-button size="small">Small Button</z-button>
        <z-button size="medium">Medium Button</z-button>
        <z-button size="large">Large Button</z-button>
        <z-button size="xlarge">Extra Large Button</z-button>
    </div>`
})

export const States = () => ({
  components: { ZButton },
  template: `<div>
        <z-button active>Active Button</z-button>
        <z-button disabled>Disabled Button</z-button>
        <z-button color="secondary" active>Active Secondary Button</z-button>
        <z-button color="secondary" disabled>Disabled Secondary Button</z-button>
    </div>`
})

export const FullWidthButton = () => ({
  components: { ZButton },
  template: `<div>
        <z-button full-width>Full Width Button</z-button>
    </div>`
})

export const DefaultLinkButton = () => ({
  components: { ZButton },
  template: `<div>
        <z-button color="link" type="link" to="https://deepsource.io/">Link Button</z-button>
    </div>`
})

export const ActiveLinkButton = () => ({
  components: { ZButton },
  template: `<div>
        <z-button color="link" type="link" to="https://deepsource.io/" active>Link Button</z-button>
    </div>`
})

export const DisabledLinkButton = () => ({
  components: { ZButton },
  template: `<div>
        <z-button color="link" type="link" to="https://deepsource.io/" disabled>Link Button</z-button>
    </div>`
})
