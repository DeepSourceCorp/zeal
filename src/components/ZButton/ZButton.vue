<template>
  <component
    :is="isLink ? 'a' : 'button'"
    :href="to || '#'"
    :disabled="disabled"
    :type="type"
    class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap"
    :class="{
      [`z-btn--${color}`]: color,
      'w-full': fullWidth,

      // Button Size & Spacing Styles
      'p-0': isLink || icon,
      'h-8 px-4 py-1 leading-loose text-xs': size == 'small' && !(isLink || iconOnly),
      'h-10 px-4 py-2 text-base leading-8': size == 'medium' && !(isLink || iconOnly),
      'h-12 px-4 py-2.5 text-lg leading-9': size == 'large' && !(isLink || iconOnly),
      'h-13 px-4 py-3 text-lg leading-9': size == 'xlarge' && !(isLink || iconOnly),

      // Set width of button when just icon is used
      'h-8 w-8': size == 'small' && iconOnly,
      'h-10 w-10': size == 'medium' && iconOnly,
      'h-12 w-12': size == 'large' && iconOnly,
      'h-13 w-13': size == 'xlarge' && iconOnly,

      // Button Base Styles
      'font-normal text-juniper': ['link', 'ghost'].includes(buttonStyle),
      'bg-juniper text-ink-400': buttonStyle == 'primary',
      'bg-ink-300 text-vanilla-100': buttonStyle == 'secondary',
      'bg-cherry text-ink-400': buttonStyle == 'danger',

      // Button Hover Styles
      'hover:underline': buttonStyle == 'link' && !disabled,
      'hover:bg-light-juniper': buttonStyle == 'primary' && !disabled,
      'hover:bg-ink-200': ['secondary', 'ghost'].includes(buttonStyle) && !disabled,
      'hover:bg-light-cherry': buttonStyle == 'danger' && !disabled,

      // Disabled Styles
      'opacity-50 cursor-not-allowed': disabled,
      'font-normal text-slate': ['link', 'ghost'].includes(buttonStyle) && disabled
    }"
  >
    <z-icon
      v-if="icon"
      :icon="icon"
      :color="iconColor"
      :size="iconSize"
      :class="{
        'mr-1': size == 'small' && $slots.default,
        'mr-1.5': size == 'medium' && $slots.default,
        'mr-2.5': size == 'large' && $slots.default,
        'mr-4': size == 'xlarge' && $slots.default
      }"
    ></z-icon>
    <slot></slot>
  </component>
</template>

<script lang="ts">
import Vue from 'vue'
import ZIcon from '@/components/ZIcon/ZIcon.vue'

export default Vue.extend({
  name: 'ZButton',
  components: {
    ZIcon
  },
  props: {
    color: {
      default: '',
      type: String
    },
    as: {
      default: 'button',
      type: String
    },
    buttonType: {
      default: 'primary',
      type: String,
      validator(val) {
        return ['primary', 'secondary', 'link', 'ghost', 'danger'].includes(val)
      }
    },
    type: {
      default: 'button',
      type: String,
      validator(val) {
        return ['submit', 'reset', 'button'].includes(val)
      }
    },
    size: {
      default: 'medium',
      type: String,
      validator(val) {
        return ['small', 'medium', 'large', 'xlarge'].includes(val)
      }
    },
    disabled: {
      default: false,
      type: Boolean
    },
    fullWidth: {
      default: false,
      type: Boolean
    },
    to: {
      default: '',
      type: String
    },
    icon: {
      type: String
    },
    iconSize: {
      type: String
    }
  },
  data() {
    return {
      defaultClasses:
        'inline-flex items-center font-medium border-transparent rounded-sm relative justify-center focus:outline-none'
    }
  },
  computed: {
    iconColor(): string {
      const colors = {
        link: 'text-juniper',
        ghost: 'text-juniper',
        primary: 'text-ink-400',
        secondary: 'text-vanilla-100',
        danger: 'text-ink-400'
      }

      return colors[this.buttonStyle as string]
    },
    iconOnly() {
      return this.icon && !this.$slots['default']
    },
    buttonStyle(): string {
      return (this.buttonType || this.color || 'primary') as string
    },
    isLink() {
      return this.as === 'link' || this.to
    }
  }
})
</script>
