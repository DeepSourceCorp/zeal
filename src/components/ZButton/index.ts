import ZButton from './ZButton.vue';

export { ZButton };
