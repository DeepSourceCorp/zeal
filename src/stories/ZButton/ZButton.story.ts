import ZButton from '../../components/ZButton/ZButton.vue'
import { Meta, Story } from '@storybook/addon-docs/blocks'
import { component } from 'vue/types/umd'

export const ButtonWithText = () => ({
  components: { ZButton },
  template: '<z-button>Hello Button</button>',
  title: 'ZButton'
})

ButtonWithText.story = {
  name: 'Button with Text'
}
