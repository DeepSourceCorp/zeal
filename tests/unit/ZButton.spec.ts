import Vue from 'vue'
import { createLocalVue, mount, Wrapper, shallowMount } from '@vue/test-utils'

import ZButton from '../../src/components/ZButton'

describe('Button component', () => {
    let mountFunction: (options?: object) => Wrapper<Vue>
    let localVue: typeof Vue
    beforeEach(() => {
        localVue = createLocalVue()
        mountFunction = (options = {}) => {
            return mount(ZButton, {
                localVue, 
                ...options
            })
        }
    })
    it('renders a button tag when the component is mounted', () => {
        expect(mountFunction().html()).toMatchSnapshot()
    })

    it('renders a button tag with the given text', () => {
        expect(mountFunction({
            slots: {
                default: 'Hello World'
            }
        }).html()).toMatchSnapshot()
    })

    it('renders a button with primary color', () => {
        expect(mountFunction({
            propsData: {
                color: 'primary'
            },
            slots: {
                default: 'Primary Button'
            }
        }).html()).toMatchSnapshot();
    })
    it('renders a button with secondary color', () => {
        expect(mountFunction({
            propsData: {
                color: 'secondary'
            },
            slots: {
                default: 'Secondary Button'
            }
        }).html()).toMatchSnapshot();
    })
    it('renders a button with the given size(large)', () => {
        expect(mountFunction({
            propsData: {
                size: 'large'
            },
            slots: {
                default: 'Large Button'
            }
        }).html()).toMatchSnapshot();
    })
    it('renders a disabled button when the disabled attribute is passed', () => {
        expect(mountFunction({
            propsData: {
                disabled: true
            },
            slots: {
                default: 'Disabled Button'
            }
        }).html()).toMatchSnapshot();
    })
    it('renders a full-width button', () => {
        expect(mountFunction({
            propsData: {
                fullWidth: true
            },
            slots: {
                default: 'Full Width Button'
            }
        }).html()).toMatchSnapshot();
    })
     it('renders a link tag when type is specified as link', () => {
        expect(mountFunction({
            propsData: {
                type: 'link',
                color: 'link',
                to: 'https://deepsource.io/'
            },
            slots: {
                default: 'Link Button'
            }
        }).html()).toMatchSnapshot();
    })
    it('should use custom classes', () => {
        const wrapper = mountFunction({
            propsData: {
                customClasses: 'bg-transparent'
            },
            slots: {
                default: 'Custom Styled Button'
            }
        });

        expect(wrapper.classes('bg-transparent')).toBe(true);
    })
    // it.only('button should emit a click event', () => {
    //     const sayHello = () => {
    //         console.log('hello')
    //     }
    //     const wrapper = mountFunction({
    //         propsData: {
    //             onclick: sayHello()
    //         }
    //     });
    //     const click = jest.fn()
    //     wrapper.vm.$on('click', click)
    //     wrapper.trigger('click')
    //     expect(click.mock.calls).toHaveLength(1)
    // })
})
