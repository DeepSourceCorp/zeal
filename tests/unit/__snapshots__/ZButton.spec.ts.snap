// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders a button tag when the component is mounted 1`] = `
<button href="#" type="button" class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap h-10 px-4 py-2 text-base leading-8 bg-juniper text-ink-400 hover:bg-light-juniper">
  <!---->
</button>
`;

exports[`Button component renders a button tag with the given text 1`] = `
<button href="#" type="button" class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap h-10 px-4 py-2 text-base leading-8 bg-juniper text-ink-400 hover:bg-light-juniper">
  <!----> Hello World
</button>
`;

exports[`Button component renders a button with primary color 1`] = `
<button href="#" type="button" class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap z-btn--primary h-10 px-4 py-2 text-base leading-8 bg-juniper text-ink-400 hover:bg-light-juniper">
  <!----> Primary Button
</button>
`;

exports[`Button component renders a button with secondary color 1`] = `
<button href="#" type="button" class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap z-btn--secondary h-10 px-4 py-2 text-base leading-8 bg-juniper text-ink-400 hover:bg-light-juniper">
  <!----> Secondary Button
</button>
`;

exports[`Button component renders a button with the given size(large) 1`] = `
<button href="#" type="button" class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap h-12 px-4 py-2.5 text-lg leading-9 bg-juniper text-ink-400 hover:bg-light-juniper">
  <!----> Large Button
</button>
`;

exports[`Button component renders a disabled button when the disabled attribute is passed 1`] = `
<button href="#" disabled="disabled" type="button" class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap h-10 px-4 py-2 text-base leading-8 bg-juniper text-ink-400 opacity-50 cursor-not-allowed">
  <!----> Disabled Button
</button>
`;

exports[`Button component renders a full-width button 1`] = `
<button href="#" type="button" class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap w-full h-10 px-4 py-2 text-base leading-8 bg-juniper text-ink-400 hover:bg-light-juniper">
  <!----> Full Width Button
</button>
`;

exports[`Button component renders a link tag when type is specified as link 1`] = `
<a href="https://deepsource.io/" type="link" class="inline-flex items-center justify-center font-medium transition-colors duration-300 ease-in-out rounded-sm focus:outline-none whitespace-nowrap z-btn--link p-0 bg-juniper text-ink-400 hover:bg-light-juniper">
  <!----> Link Button
</a>
`;
